bin/pp_ast.ml

```diff
@@ -0,0 +1,219 @@
+open Ppxlib
+
+module Kind = struct
+  type t = Signature | Structure | Expression | Pattern | Core_type
+
+  let to_utils_kind = function
+    | Structure -> Ppxlib_private.Utils.Kind.Impl
+    | Signature -> Ppxlib_private.Utils.Kind.Intf
+    | _ -> assert false
+end
+
+module Ast = struct
+  type t =
+    | Str of structure
+    | Sig of signature
+    | Exp of expression
+    | Pat of pattern
+    | Typ of core_type
+end
+
+module Input = struct
+  type t = Stdin | File of string | Source of string
+
+  let to_lexbuf t =
+    let all_source =
+      match t with
+      | Stdin -> Stdppx.In_channel.input_all stdin
+      | File fn -> Stdppx.In_channel.(with_file fn ~f:input_all)
+      | Source s -> s
+    in
+    Lexing.from_string all_source
+
+  let from_string = function
+    | "-" -> Stdin
+    | s when Sys.file_exists s -> File s
```

Comment:
Do we have something lying about for doing `file_exists` checks that isn't this? IIRC `Sys.file_exists` converts _any_ error into `false` (e.g. a permission error) which is confusing for an end user maybe ? If we don't have anything easily at hand then this is fine.

Reply:
We might have something in `stdppx`, I'll take a look!

Reply:
From a quick look it doesn't seem like we have something satisfying lying around, we're already using `Sys.file_exists` in a few places.

I suggest we leave it at that for this PR but we can eventually add a better version to `stdppx` later on if needed.